### PR TITLE
Add comments documenting ownership semantics

### PR DESCRIPTION
**Pull request by dgquintas**

### CODE REVIEW

File: src/core/ext/client_config/resolver_result.h [hunk 1]
```diff
@@ -57,7 +57,7 @@ void grpc_resolver_result_unref(grpc_exec_ctx* exec_ctx,
                                 grpc_resolver_result* result);
 
 /// Caller does NOT take ownership of result.
```
Comment (dgquintas) [inline src/core/ext/client_config/resolver_result.h]: This is what I'm not too happy about, returning something that's not meant to be owned as a mutable.

**Comment by markdroth** [comment]
As per our conversation, unless we have a strong convention in core that `const char*` always points to a statically allocated string, then I personally prefer the code as-is.  The casting when calling `gpr_free()` is annoying but not conceptually problematic to me.

**Comment by dgquintas** [comment]
PTAL. I may have found a good compromise.

**Comment by dgquintas** [comment]
To summarize: my intent is to uphold the convention that const pointers aren't owned and shouldn't be freed. Where a cast is required due to the particularities of the code, a comment has been added.

- The signature of `grpc_resolver_result_get_server_name` keeps the `const char*` signaling that the returned value is NOT owned by the caller.
- The relevant `gpr_free`s don't need a cast that would break the const-ownership convention.

File: src/core/ext/lb_policy/pick_first/pick_first.c [hunk 2]
```diff
@@ -466,7 +466,9 @@ static grpc_lb_policy *create_pick_first(grpc_exec_ctx *exec_ctx,
     }
 
     memset(&sc_args, 0, sizeof(grpc_subchannel_args));
-    sc_args.server_name = args->server_name;
+    /* The following cast is fine because server_name argument will be copied as
```
Comment (markdroth) [inline src/core/ext/lb_policy/pick_first/pick_first.c]: The one part of this PR that I do like is adding a comment here (and in the corresponding place in round_robin.c) that indicates that the value will be copied as part of subchannel creation.  Even though I'm not in favor of the const changes, these comments seem valuable.

Comment (dgquintas) [inline src/core/ext/lb_policy/pick_first/pick_first.c]: Done.

**Comment by dgquintas** [comment]
Thank you @markdroth for that very comprehensive comment. I stand convinced :) I've reverted the changes and added a single commit for the addition of the comments. PTAL!

**Comment by ctiller** [comment]
Removing myself from reviewers unless arbitration is required :)

**Pull request merged**